File: extension.ts
//% color="#359eff" weight=20 icon="\uf1b2"
namespace Cube {
    let BUS_SERVO_ENABLE=false;
    let Color_Recognize:number;
    export enum GPIO_ID{
        A0,
        A1,
        B0,
        B1,
        C0,
        C1,
        D0,
        D1,
        E0,
        E1,
        F0,
        F1,
        G0,
        G1,
        H0,
        H1
    }
    export enum Analog_Pins{
        A0,
        A1,
        B0,
        B1,
        C0,
        C1,
        D0,
        D1
    }
    export enum PORT_ID{
        //% block="端口A"
        PORTA,
        //% block="端口B"
        PORTB,
        //% block="端口C"
        PORTC,
        //% block="端口D"
        PORTD
    }
    export enum Pin_MODE{
        //% block="输入"
        INPUT_NOPULL,
        //% block="上拉输入"
        INPUT_PULLUP,
        //% block="下拉输入"
        INPUT_PULLDOWN,
        //% block="数字输出"
        OUTPUT,
        //% block="模拟输出"
        PWM,
        //% block="模拟输入(仅支持模拟引脚)"
        ADC
    }
    export enum Pin_Level{
        //% block="高"
        High,
        //% block="低"
        Low
    }
    export enum Sensor_ID{
        //% block="按钮模块"
        Button,
        //% block="触摸模块"
        Touch,
        //% block="手势&颜色模块"
        Gesture,
        //% block="红外光电模块"
        IR,
        //% block="超声波测距模块"
        Ultrasonar,
    }
    export enum Motor_ID{
        M1,
        M2,
        M3,
        M4
    }
    export enum Motor_Dir{
        //% block="正转"
        Forward=1,
        //% block="反转"
        Backward=2,
        //% block="刹车"
        Brake=0
    }
    export enum Move_base_dir{
        //% block="前进"
        Forward=1,
        //% block="后退"
        Backward=2,
        //% block="左转"
        Turn_left=3,
        //% block="右转"
        Turn_right=4,
        //% block="刹车"
        Brake=0
    }

    export enum IMU_AXIS{
        Yaw,
        Roll,
        Pitch
    }

    export enum End_TYPE{
        //% block="左转路口"
        Left=2,
        //% block="右转路口"
        Right=3,
        //% block="左转或右转路口"
        Cross=1,
    }

    export enum Sucker_Operation{
        //% block="吸起"
        Hold,
        //% block="释放"
        Release
    }

    export enum Go_Distance{
        //% block="前进"
        Forward,
        //% block="后退"
        Backward
    }

    export enum Is_Wait{
        //% block="是"
        Blocking=1,
        //% block="否"
        NonBlocking=0
    }

    //% block="复位编程盒" advanced=true
    //% shim=Cube::Init
    export function Init() {
        return
    }

    //% block="设置引脚模式%id|%mode"
    //% shim=Cube::Set_Pin_Mode
    export function Set_Pin_Mode(id:GPIO_ID,mode:Pin_MODE){
        return
    }
    //% block="数字输出 引脚%id 设为%level"
    //% shim=Cube::Set_Pin_Value
    export function Set_Pin_Value(id:GPIO_ID,level:Pin_Level){
        return
    }
    //% block="模拟输出 引脚%id 设为%pwm"
    //% shim=Cube::Set_Pin_PWM
    export function Set_Pin_PWM(id:GPIO_ID,pwm:number){
        return
    }
    //% block="数字读取 引脚%id"
    //% shim=Cube::Get_Pin_Value
    export function Get_Pin_Value(id:GPIO_ID){
        return 0
    }
    //% block="模拟读取 引脚%id"
    //% shim=Cube::Get_ADC_Value
    export function Get_ADC_Value(id:Analog_Pins){
        return 0
    }
    //% block="初始化%port|为%sensor"
    //% shim=Cube::Init_Port
    export function init_port(id:PORT_ID,sensor:Sensor_ID){
        return
    }
    //% block="设置马达%id|模式%dir|功率%pwm"
    //% shim=Cube::Motor pwm.defl=0 pwm.min=0 pwm.max=255
    export function Motor(id:Motor_ID,dir:Motor_Dir,pwm:number){
        return
    }
    //% block="读取陀螺仪数据%dir"
    //% shim=Cube::Get_Imu 
    //% advanced=true
    export function Get_Imu(dir:IMU_AXIS){
        return 0
    }

    //% block="初始化视觉模块"
    //% group="综合技能"
    export function Init_MU(){
        serial.redirect(
        SerialPin.P8,
        SerialPin.P12,
        BaudRate.BaudRate9600
        )
        basic.pause(100)
        MUVisionSensor.begin(MUVisionSensor.SENSORS.MU00, MUVisionSensor.PORT.Serial)
        MUVisionSensor.VisionBegin(MUVisionSensor.SENSORS.MU00, MUVisionSensor.ENABLES.enable, MUVisionSensor.VISION_TYPE.VISION_COLOR_RECOGNITION)
        MUVisionSensor.set_WB(MUVisionSensor.SENSORS.MU00, MUVisionSensor.WBMODE.WB_LOCK)
        music.playTone(523, music.beat(BeatFraction.Whole))
        music.playTone(784, music.beat(BeatFraction.Whole))
    }

    //% block="自动标定循线传感器"
    //% shim=Cube::Init_sensor 
    //% group="综合技能"
    export function Init_sensor(){
        basic.pause(3000)
        return 
    }

    //% block="循线到%end_type||是否等待到达%mode速度%speed"
    //% shim=Cube::follow_line 
    //% group="综合技能"
    //% expandableArgumentMode="toggle"
    //% mode.defl=1
    //% speed.defl=100 speed.max=100 speed.min=70
    export function follow_line(end_type:End_TYPE,mode?:Is_Wait,speed?:number){
        return 
    }
    
    //% block="循线到第%count个路口|速度%speed||是否等待到达%mode"
    //% shim=Cube::follow_line_n 
    //% group="综合技能"
    //% expandableArgumentMode="toggle"
    //% count.defl=3
    //% speed.defl=100 speed.max=100 speed.min=70
    export function follow_line_n(count:number,speed:number,mode?:Is_Wait){
        return 
    }

    //% block="循线完成"
    //% shim=Cube::is_arrive_end
    //% group="综合技能"
    export function is_arrive_end():boolean{
        return true
    }
    
    /**
     * 立即停止循线（仅在非等待模式下有效）
     */
    //% block="停止循线"
    //% shim=Cube::break_follow
    //% group="综合技能"
    export function break_follow(){
        return 
    }

    /**
     * 
     * @param angle 旋转的角度，顺时针为正，逆时针为负
     */
    //% block="原地转向%angle°"
    //% shim=Cube::turn_angle
    //% angle.defl=90
    //% group="综合技能"
    export function turn_angle(angle: number){
        return 
    }    

    /**
     * 
    * @param angle 旋转到指定角度，以开机时车头正对的角度为0°，顺时针为正，设置范围为-180~180
     */
    //% block="原地转向到面向%angle°"
    //% shim=Cube::turn_to_angle
    //% angle.defl=90
    //% group="综合技能"
    export function turn_to_angle(angle: number){
        return 
    } 

    //% block="%dir|距离%dist mm||速度%speed"
    //% shim=Cube::go_distance
    //% dist.defl=200
    //% group="综合技能"
    //% expandableArgumentMode="toggle"
    //% speed.defl=200 speed.max=600 speed.min=100
    export function go_distance(dir:Go_Distance, dist: number, speed?:number){
        return 
    }

    //% block="%op目标物体"
    //% shim=Cube::suck
    //% group="综合技能"
    export function suck(op:Sucker_Operation){
        return 
    }
    export enum object_color{
        //% block="蓝色"
        Blue=1,
        //% block="黄色"
        Yellow=2,
        //% block="红色"
        Red=3,
        //% block="绿色"
        Green=4,
        //% block="无"
        None=0
    }
    //% block="启动颜色检测"
    //% group="综合技能"
    export function Object_Detect(){
        let yellow = 0
        let blue = 0
        let red = 0
        let green = 0
        Color_Recognize=0
        while (!(Cube.is_arrive_end())) {
            if (MUVisionSensor.MuVs2GetColorRCGLabel(MUVisionSensor.SENSORS.MU00, 50, 100)) {
                let color = MUVisionSensor.get_color_value(MUVisionSensor.SENSORS.MU00, MUVisionSensor.ColorParams.LABLE)
                switch(color){
                    case MUVisionSensor.COLOR_TYPE.BLUE:blue++;break;
                    case MUVisionSensor.COLOR_TYPE.YELLOW:yellow++;break;
                    case MUVisionSensor.COLOR_TYPE.RED:red++;break;
                    case MUVisionSensor.COLOR_TYPE.GREEN:green++;break;
                }
            }
            basic.pause(50)
        }
        
        if (blue > yellow&&blue>red&&blue>green) {
            Color_Recognize=1
        } else if (yellow > blue&&yellow>red&&yellow>green) {
            Color_Recognize=2
        } else if (red > blue&&red>yellow&&red>green) {
            Color_Recognize=3
        } else if (green > blue&&green>red&&green>yellow) {
            Color_Recognize=4
        } else {
            Color_Recognize=0
        }
    }
    //% block="检测到物体颜色为%color"
    //% group="综合技能"
    export function Get_Color(color:object_color){
        if(Color_Recognize==color)
            return true
        else
            return false
    }
    
    //% block="总线舵机控制|ID %ID|角度 %value|时间 %time ms"
    //% time.defl=500 time.min=0
    //% value.min=0 value.max=180
    //% num.fieldEditor="gridpicker" num.fieldOptions.columns=4
    //% advanced=true
    export function bus_Servo(ID: number, value: number, time: number): void {
        if(!BUS_SERVO_ENABLE){
            serial.redirect(SerialPin.P8,SerialPin.P12,115200);
            BUS_SERVO_ENABLE=true;
        }
        serial.writeString("#");
        if(ID<10)
            serial.writeString("00");
        else if(ID<100)
            serial.writeString("0");
        serial.writeNumber(ID);
        serial.writeString("P");
        let pwm=Math.map(value,0,180,500,2500);
        pwm=Math.round(pwm);
        if(pwm<1000){
            serial.writeString("0");
        }
        serial.writeNumber(pwm);
        serial.writeString("T");
        if(time<1000){
            serial.writeString("0");
        }
        serial.writeNumber(time);
        serial.writeString("!");
    }
    //% block="初始化底盘 左马达%left|右马达%right"
    //% shim=Cube::Set_move_base group="底盘控制"
    export function Set_move_base(left:Motor_ID,right:Motor_ID){
        return
    }
    //% block="底盘控制 %dir| 速度%speed"
    //% shim=Cube::move_base group="底盘控制"
    export function move_base(dir:Move_base_dir, speed:number){
        return
    }
    //% block="设置 左马达%left| 右马达%right 速度（±255）"
    //% left.min=-255 left.max=255 right.min=-255 right.max=255
    //% shim=Cube::move_motor group="底盘控制"
    export function move(left:number, right:number){
        return
    }
    //% block="设置 左轮%left| 右轮%right 速度mm/s（±800）"
    //% left.min=-800 left.max=800 right.min=-800 right.max=800
    //% shim=Cube::move_motor_close group="底盘控制"
    export function move_close(left:number, right:number){
        return
    }
    //% block="获取测距传感器数据 编号%begin|至%end"
    //% advanced=true
    export function Get_VL53L0X(begin:number,end:number){
        let data=[0]
        _update_vl53l0x(begin,end);
        for (let index = 0; index < end-begin+1; index++) {
            data.push(_get_vl53l0x(index))
        }
        data.shift()
        return data
    }

    //% shim=Cube::Update_VL53L0X
    export function _update_vl53l0x(begin:number,end:number){
        return
    }
    //% shim=Cube::Get_VL53L0X
    export function _get_vl53l0x(index:number){
        return 0
    }

    //% shim=Cube::test
    export function test(){
        return 0
    }
}

File: extension.cpp
#include "pxt.h"
#include "i2c_hw_interface.h"
using namespace pxt;
typedef union{
	struct{
		uint8_t D_pin:3;
		uint8_t A_pin:3;
	};
	uint8_t Port_reg_value;
}Port_Reg;
typedef union{
	struct{
		uint8_t PORT_A:4;
		uint8_t PORT_B:4;
	};
	struct{
		uint8_t PORT_C:4;
		uint8_t PORT_D:4;
	};
	uint8_t Port_reg_value;
}Device_Reg;


namespace Cube {
    CubeI2C* i2c=new CubeI2C(&uBit.i2c,0x70);
    uint8_t Left_wheel=255;
    uint8_t Right_wheel=255;
    uint8_t data[16]={0};
    void wait_for_cmd_finish(){
        uint8_t is_finish;
        while(1){
            wait_ms(100);
            i2c->I2CRead(0x80,&is_finish);
            if(!is_finish)
                break;
        }
    }
    //%
    void Init() {
        uint8_t temp;
        uint8_t err_cnt=0;
        while(i2c->I2CRead(0x01,&temp)!=0)
        {
            if(err_cnt++>20)
            {
                uBit.display.scroll("init error");
                uBit.reset();
            }
            wait_ms(10);
        }
        // while(i2c->I2CRead(0x02,&temp)==0);
        // i2c->I2CWrite(0x03,1);
    }
    //%
    int test() {
        uint8_t temp;
        return i2c->I2CRead(0x01,&temp);
    }
    //%
    void Motor(int id,int dir,int pwm){        
        i2c->I2CWrite(0x34+id,dir,pwm);
    }
    //%
    void Set_move_base(int left,int right){
        Left_wheel=left;
        Right_wheel=right;
        Motor(Left_wheel,0,0);
        Motor(Right_wheel,0,0);
    }
    //%
    void move_base(int dir, int speed){
        if(Left_wheel<4&&Right_wheel<4){
            switch(dir){
                case 1:
                    Motor(Left_wheel,1,speed);
                    Motor(Right_wheel,1,speed);
                    break;
                case 2:
                    Motor(Left_wheel,2,speed);
                    Motor(Right_wheel,2,speed);
                    break;
                case 3:
                    Motor(Left_wheel,2,speed);
                    Motor(Right_wheel,1,speed);
                    break;
                case 4:
                    Motor(Left_wheel,1,speed);
                    Motor(Right_wheel,2,speed);
                    break;
                case 0:
                    Motor(Left_wheel,0,speed);
                    Motor(Right_wheel,0,speed);
                    break;
            }
        }
    }
    //%
    void move_motor(int left,int right){
        uint8_t left_dir=left>0?1:2;
        uint8_t right_dir=right>0?1:2;
        Motor(Left_wheel,left_dir,abs(left));
        Motor(Right_wheel,right_dir,abs(right));
    }
    //%
    int Get_Imu(int dir){
        uint8_t data_buf[2];
        int16_t data=0;
        i2c->I2CRead2Byte(0x10+dir*2,data_buf);
        data=data_buf[0];
        data=(data<<8)|data_buf[1];
        return data/100;
    }
    //%
    void Set_Pin_Mode(int id,int mode){
        Port_Reg port;
        i2c->I2CRead(0x05+id/2,&port.Port_reg_value);
        if(id%2==0)
        {
            if(mode!=port.D_pin)
            {
                port.D_pin=mode;
                i2c->I2CWrite(0x05+id/2,port.Port_reg_value);
            }    
        }
        else
        {
            if((mode<<3)!=port.A_pin)
            {
                port.A_pin=mode;
                i2c->I2CWrite(0x05+id/2,port.Port_reg_value);
            }
        }
        wait_us(25);    
    }
    //%
    void Set_Pin_Value(int id,int value){
        uint8_t temp;
        i2c->I2CRead(0x09,&temp);
        if(value!=(temp&0x01<<id))
        {
            temp=temp^(0x01<<id);
            i2c->I2CWrite(0x09,temp);
        }    
    }
    //%
    void Set_Pin_PWM(int id,int pwm){
        i2c->I2CWrite(0x20+id,pwm);    
    }
    //%
    int Get_Pin_Value(int id){
        uint8_t temp;
        while(i2c->I2CRead(0x09,&temp)!=0);
        return temp>>id&0x01;
    }
    //%
    int Get_ADC_Value(int id){
        uint8_t temp[2]={0};
        uint16_t adc=0;
        uint8_t ret = i2c->I2CRead2Byte(0x16+id,temp);
        if(ret==0)
        {
            adc|=(uint16_t)temp[1];
            adc|=(uint16_t)(temp[0]<<8);
            return adc; 
        }
        else
            return ret;
        
    }
    //%
    void Init_Port(int port_id,int sensor_id){
        Device_Reg reg;
        if(port_id<2){
            i2c->I2CRead(0x28,&reg.Port_reg_value);
            if(port_id==0){
                if(sensor_id!=reg.PORT_A){
                    reg.PORT_A=sensor_id;
                    i2c->I2CWrite(0x28,reg.Port_reg_value);
                }
            }
            else{
                if(sensor_id!=reg.PORT_B){
                    reg.PORT_B=sensor_id;
                    i2c->I2CWrite(0x28,reg.Port_reg_value);
                }
            }
        }
        else{
            i2c->I2CRead(0x29,&reg.Port_reg_value);
            if(port_id==2){
                if(sensor_id!=reg.PORT_C){
                    reg.PORT_C=sensor_id;
                    i2c->I2CWrite(0x29,reg.Port_reg_value);
                }
            }
            else{
                if(sensor_id!=reg.PORT_D){
                    reg.PORT_D=sensor_id;
                    i2c->I2CWrite(0x29,reg.Port_reg_value);               
                }
            }
        }
    }
    //%
    void Update_VL53L0X(int begin, int end){
        uint8_t cmd=0;
        uint8_t cnt=end-begin+1;
        cmd=(0x01<<begin)|(0x01<<end);
        i2c->I2CWrite(0x41,cmd);
        while(i2c->I2CRead(data, 2*cnt));
        
    }

    //%
    uint16_t Get_VL53L0X(int index){
        uint16_t ret=data[index*2];
        ret=(ret<<8)|data[index*2+1];
        return ret;
    }

    //%
    void Init_sensor(){
        i2c->I2CWrite(0x50,0,0);
        wait_for_cmd_finish();
    }

    //%
    void follow_line(int end_type,int is_wait){
        i2c->I2CWrite(0x51,end_type,0);
        if(is_wait)
            wait_for_cmd_finish();
    }

    //%
    void turn_angle(int angle){
        i2c->I2CWrite(0x52,(uint8_t)(angle>>8),(uint8_t)(angle&0x00ff));
        wait_for_cmd_finish();
    }

    //%
    void go_distance(int dist){
        i2c->I2CWrite(0x53,(uint8_t)(dist>>8),(uint8_t)(dist&0x00ff));
        wait_for_cmd_finish();
    }

    //%
    void break_follow(){
        i2c->I2CWrite(0x54,0,0);
    }
    void release(){
        i2c->I2CWrite(0x56,0,0);
        wait_ms(200);
    }
    //%
    void suck(int op){
        if(op==0)
        {
            int pitch0 = Get_Imu(2);
            Motor(3,2,255);//吸盘下降
            wait_ms(50);
            while(Get_Imu(2)-pitch0<2||Get_Imu(2)-pitch0>-2)
            {
                wait_ms(50);
            }
            Motor(3,0,255);
            i2c->I2CWrite(0x55,0,0);
            wait_ms(100);
            Motor(3,1,255);
            wait_ms(500);
            Motor(3,0,0);
        }    
        else if(op==1)
            release();
    }

    

    //%
    int is_arrive_end(){
        uint8_t is_finish;
        i2c->I2CRead(0x80,&is_finish);
            return !is_finish;
    }
}

